lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -818,6 +828,10 @@ matchWitness(TypeChecker &tc,
     // Check that the mutating bit is ok.
     if (checkMutating(funcReq, funcWitness, funcWitness))
       return RequirementMatch(witness, MatchKind::MutatingConflict);
+    if (funcWitness->isNonMutating() && funcReq->isConsuming())
```

Comment:
This one has the opposite symptoms but the same problem where we thunk to guaranteed and regress performance.